dashboard/test/controllers/api/v1/amazon_future_engineer_controller_test.rb

```diff
@@ -2,46 +2,94 @@
 
 class Api::V1::AmazonFutureEngineerControllerTest < ActionDispatch::IntegrationTest
   setup do
-    @teacher = create :teacher
+    CDO.stubs(:afe_pardot_form_handler_url).returns('fake_pardot_url.com')
   end
 
   test 'logged out cannot submit' do
-    # Add expectation that there are no HTTP requests to Pardot.
+    Net::HTTP.expects(:post_form).never
 
     post '/dashboardapi/v1/amazon_future_engineer_submit',
-      params: {
-        'first-name' => 'test',
-        'last-name' => 'test'
-      }, as: :json
+      params: valid_params, as: :json
 
     assert_response :forbidden
   end
 
-  test 'logged in can submit' do
+  test 'submit returns BAD REQUEST when params are malformed' do
+    Net::HTTP.expects(:post_form).never
+
+    # Intentionally missing required field traffic-source
+    sign_in create :teacher
+    post '/dashboardapi/v1/amazon_future_engineer_submit',
+      params: valid_params.delete('traffic-source'), as: :json
+
+    assert_response :bad_request, "Expected BAD REQUEST, got: #{response.status}\n#{response.body}"
+  end
+
+  test 'submit returns BAD REQUEST when params are missing' do
     Net::HTTP.stubs(:post_form)
-    Net::HTTP.expects(:post_form).returns(FakeResponse.new)
 
-    sign_in(@teacher)
+    sign_in create :teacher
+    post '/dashboardapi/v1/amazon_future_engineer_submit'
 
-    post '/dashboardapi/v1/amazon_future_engineer_submit',
-      params: {
-        'first-name' => 'test',
-        'last-name' => 'test',
-        'email' => 'test@code.org',
-        'nces-id' => '123456789012',
-        'street-1' => 'test street',
-        'city' => 'seattle',
-        'state' => 'wa',
-        'zip' => '98105',
-        'inspirational-marketing-kit' => '0',
-        'csta-plus' => '0',
-        'aws-educate' => '0',
-        'amazon-terms' => '1',
-        'new-code-account' => '0',
-        'registration-date-time' => '1997-07-16T19:20:30+01:00'
-      }, as: :json
-
-    assert_response :success, "Failed response: #{response.body}"
+    assert_response :bad_request, "Expected BAD REQUEST, got: #{response.status}\n#{response.body}"
```

Comment:
One note is we might actually want `response.code` (assuming the `FakeResponse` is right)?

Reply:
NVM, tested and status works as well. A little confused by [docs here](https://api.rubyonrails.org/classes/ActionDispatch/Response.html#method-i-code), does `status` represent a way to set the status of a response, or get it, or both?

Reply:
Looking at those docs: `status` is both a readonly attribute on a response object, documented here:

![image](https://user-images.githubusercontent.com/1615761/85790533-328ba500-b6e5-11ea-8828-3cc29a5d4742.png)

...and also the name of a method that can be used to set the `status` attribute, documented here:

![image](https://user-images.githubusercontent.com/1615761/85790590-4e8f4680-b6e5-11ea-8b24-4c58279c9441.png)

Finally, it looks like the `code` method is only different from `status` in that it always returns a string, where status codes are normally numbers.

![image](https://user-images.githubusercontent.com/1615761/85790674-78486d80-b6e5-11ea-9d7c-c6aab6d5eabc.png)